### PR TITLE
[Export] Introudce Epilogue for buffer mutation

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94147

See comment inline.

### CODE REVIEW

File: export_workspace/example_mutate_buffer.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        self._bin_num_examples.index_add_(
+            dim=0, index=index, source=supervision_weight.squeeze()
+        )  # ScatterWeightedSum
+
+        return read
+
+    inp = (torch.tensor(0.5, dtype=torch.double), torch.tensor(2, dtype=torch.int), torch.rand(2,2, requires_grad=True))
+
+aten = torch.ops.aten
+my_decomp = get_decompositions(
```
Comment (SherlockNoMad) [inline export_workspace/example_mutate_buffer.py:41]: Both index_add_ and index_copy_ can be decomposed into index_put_.

File: export_workspace/example_mutate_buffer.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+        # Buffer update
+        _weighted_slices = torch.index_select(self._bin_num_examples, 0, index) * 1.0
+
+        self._bin_num_examples.index_copy_(
+            dim=0, index=index.long(), source=_weighted_slices.squeeze()
+        )
+        self._bin_num_examples.index_add_(
+            dim=0, index=index, source=supervision_weight.squeeze()
+        )  # ScatterWeightedSum
```
Comment (SherlockNoMad) [inline export_workspace/example_mutate_buffer.py:34]: Notice that these are two inplace ops mutating the same buffer. 
If we were to preserve them in export, computes in the Epilogue needs to be ordered.

File: export_workspace/example_mutate_buffer.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+        # File: /scratch/bahuang/work/repos/pytorch/export_workspace/example_mutate_buffer.py:30, code: dim=0, index=index.long(), source=_weighted_slices.squeeze()
+        _to_copy_default = torch.ops.aten._to_copy.default(arg1, dtype = torch.int64)
+        squeeze_default = torch.ops.aten.squeeze.default(mul_tensor);  mul_tensor = None
+
+        # No stacktrace found for following nodes
+        _tensor_constant0_2 = self._tensor_constant0
+
+        # File: /scratch/bahuang/work/repos/pytorch/export_workspace/example_mutate_buffer.py:29, code: self._bin_num_examples.index_copy_(
+        index_put__default = torch.ops.aten.index_put_.default(_tensor_constant0_2, [_to_copy_default], squeeze_default);  _tensor_constant0_2 = _to_copy_default = squeeze_default = None
```
Comment (SherlockNoMad) [inline export_workspace/example_mutate_buffer.py:78]: index_copy_ decomposed into index_put_(..., accumulate=False)

File: export_workspace/example_mutate_buffer.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+        # File: /scratch/bahuang/work/repos/pytorch/export_workspace/example_mutate_buffer.py:33, code: dim=0, index=index, source=supervision_weight.squeeze()
+        squeeze_default_1 = torch.ops.aten.squeeze.default(arg0);  arg0 = None
+
+        # File: /scratch/bahuang/work/repos/pytorch/export_workspace/example_mutate_buffer.py:32, code: self._bin_num_examples.index_add_(
+        index_put__default_1 = torch.ops.aten.index_put_.default(index_put__default, [arg1], squeeze_default_1, True);  index_put__default = arg1 = squeeze_default_1 = None
```
Comment (SherlockNoMad) [inline export_workspace/example_mutate_buffer.py:84]: index_add_ decomposed into index_put_(..., accumulate=True)

File: torch/_export/logical_schema.py [hunk 5]
```diff
@@ -269,6 +269,18 @@ class Graph:
     symint_values: List[SymIntValue]
 
 
+@dataclass(init=False)
+class Epilogue:
+    inputs: List[TensorArgument]
```
Comment (SherlockNoMad) [inline torch/_export/logical_schema.py:274]: For the example above, 

inputs should have [index_for_first_index_put, value_for_firs_index_put, index_for_second_index_put, value_for_second_index_put]

File: torch/_export/logical_schema.py [hunk 6]
```diff
@@ -269,6 +269,18 @@ class Graph:
...
+
+    # Buffers and parameters that was mutated by the epilogue program
+    mutated_values: List[TensorArgument]
```
Comment (SherlockNoMad) [inline torch/_export/logical_schema.py:277]: mutated_values should be [_tensor_constant0]

File: torch/_export/logical_schema.py [hunk 7]
```diff
@@ -269,6 +269,18 @@ class Graph:
...
+
+    # Epiloge computes to mutate the parameters and buffers of the GraphModule
+    # Nodes MUST be 'aten.copy_' or 'aten.index_put_'
+    nodes: List[Node]
```
Comment (SherlockNoMad) [inline torch/_export/logical_schema.py:281]: nodes should contain two aten.index_put_ nodes.

File: torch/_export/logical_schema.py [hunk 8]
```diff
@@ -279,6 +291,8 @@ class GraphModule:
 
     graph: Graph    # Only one Graph per GraphModule
 
+    epilogue: Epilogue   # Epilogue program for the mutation of parameters and buffers
```
Comment (suo) [inline torch/_export/logical_schema.py:294]: nit: this should contain the name `Mutation` to make clear its purpose.

**Review by suo — CHANGES_REQUESTED** [review]
pedantic complaint about schema, otherwise lgtm

File: torch/_export/logical_schema.py [hunk 9]
```diff
@@ -269,6 +269,18 @@ class Graph:
     symint_values: List[SymIntValue]
 
 
+@dataclass(init=False)
+class Epilogue:
```
Comment (zhxchen17) [inline torch/_export/logical_schema.py:273]: an alternative design I think also works here:
```Python
class Epilogue:
    copies: List[...]
    index_puts: List[...]
```
and inputs data can be expressed implicitly inside `...` part.

**Pull request closed without merging**